internal/enricher/geolocation.go

```diff
@@ -0,0 +1,185 @@
+	GetDB(string, string) error
+}
+
+type Geolocation struct {
+	City     string `json:"city"`
+	Country  string `json:"country"`
+	Region   string `json:"region"`
+	Postal   string `json:"postal"`
+	Location string `json:"location"`
+	Timezone string `json:"timezone"`
+}
+
+type geoEnricher struct {
+	fetcher geolocation.GeoFetcher
+	logger  logger.Logger
+	stats   stats.Stats
+}
+
+func NewGeoEnricher(
+	dbProvider GeoDBProvider,
+	config *config.Config,
+	log logger.Logger,
+	statClient stats.Stats,
+) (PipelineEnricher, error) {
+	upstreamDBKey := config.GetString("Geolocation.db.key", "geolite2City.mmdb")
+	downloadPath := config.GetString("Geolocation.db.downloadPath", "geolite2City.mmdb")
+
+	err := dbProvider.GetDB(upstreamDBKey, downloadPath)
+	if err != nil {
+		return nil, fmt.Errorf("getting db from upstream: %w", err)
+	}
+
+	fetcher, err := geolocation.NewMaxmindDBReader(downloadPath)
+	if err != nil {
+		return nil, fmt.Errorf("creating new instance of maxmind's geolocation fetcher: %w", err)
+	}
+
+	return &geoEnricher{
+		fetcher: fetcher,
+		stats:   statClient,
+		logger:  log.Child("geolocation"),
+	}, nil
+}
+
+// Enrich function runs on a request of GatewayBatchRequest which contains
+// multiple singular events from a source. The enrich function augments the
+// geolocation information per event based on IP address.
+func (e *geoEnricher) Enrich(sourceId string, request *types.GatewayBatchRequest) error {
+	e.logger.Debugw("received a call to enrich gateway events for source", "sourceId", sourceId)
+
+	if request.RequestIP == "" {
+		e.stats.NewTaggedStat("proc_geo_enrincher_empty_ip", stats.CountType, stats.Tags{
+			"sourceId": sourceId,
+		}).Increment()
+
+		return nil
+	}
+
+	defer func(from time.Time) {
+		e.stats.NewTaggedStat(
+			"proc_geo_enricher_request_latency",
+			stats.TimerType,
+			stats.Tags{"sourceId": sourceId}).Since(from)
+	}(time.Now())
+
+	geoip, err := e.fetcher.GeoIP(request.RequestIP)
+	if err != nil {
+		e.stats.NewTaggedStat(
+			"proc_geo_enricher_geoip_lookup_failed",
+			stats.CountType,
+			stats.Tags{
+				"sourceId": sourceId,
+				"validIP":  strconv.FormatBool(errors.Is(err, geolocation.ErrInvalidIP)),
+			}).Increment()
+
+		return fmt.Errorf("unable to enrich the request with geolocation: %w", err)
+	}
+
+	// for every event if we have context object set
+	// only then we add to geo section
+	for _, event := range request.Batch {
+		if context, ok := event["context"].(map[string]interface{}); ok {
+			context["geo"] = extractGeolocationData(geoip)
+		}
+	}
+
+	return nil
+}
+
+type NoOpGeoEnricher struct{}
+
+func (e NoOpGeoEnricher) Enrich(sourceId string, request *types.GatewayBatchRequest) error {
+	return nil
+}
+
+type geoDBProviderImpl struct {
+	s3Manager filemanager.FileManager
+}
+
+// GetDB simply fetches the database from the upstream located at the key defined
+// in the argument.
+func (p *geoDBProviderImpl) GetDB(key, downloadPath string) error {
```

Comment:
Should we pass context here? 
```suggestion
func (p *geoDBProviderImpl) GetDB(ctx context.Context, key, downloadPath string) error {
```